Dictionary values saved for instanced scene only when different from parent

pouleyKetchoupp opened the pull request:
This change makes dictionary comparison operators to check for differences in data, following the same logic as for arrays.

It allows dictionary export variables in instanced scenes to be compared to parents, so they are saved only when the data has been modified for the instance.

Fixes #29221

pouleyKetchoupp committed:
Fix dictionary values saved for instanced scene only when different from parent

bojidar-bg commented on core/variant_op.cpp:
```diff
@@ -430,10 +430,13 @@ void Variant::evaluate(const Operator &p_op, const Variant &p_a,
+				if (dict_a->size() != dict_b->size())
+					_RETURN(false);
+
+				_RETURN(dict_a->hash() == dict_b->hash());
```
I think it might be better to compare dictionaries similar to how arrays are compared below: element-by-element.

pouleyKetchoupp replied:
Ok, I can do that. I will also sort the keys so the comparison doesn't depend on the insertion order. Is that ok?

akien-mga replied:
Why isn't hash comparison enough though? Isn't it more efficient this way if we only care about equality/inequality?

bojidar-bg replied:
Well, calculating the hash also goes through all the entries, so it is the same.

> I will also sort the keys so the comparison doesn't depend on the insertion order. Is that ok?

No need for sorting, as long as the size matches. If `a.size() == b.size()` and for all `x`, `x` from `a`, `b.has(x) && a[x] == b[x]`, then all keys/values in `a` are also present in `b` and vice-versa.

pouleyKetchoupp replied:
Yep, makes sense for the sorting.

I was thinking I could even make it a bit more efficient with a new method `get_entries` in Dictionary to get a list of key/value pairs directly from `a` without accessing `a[x]` for each key, and then use `b.getptr(x)` to check it has the key and get the value for comparison. Or is it too much?

bojidar-bg replied:
I guess you can do that as well, though it seems a bit complicated.

pouleyKetchoupp replied:
Ok, I ended up doing it the optimal way, it wasn't too difficult.

akien-mga commented:
Some comments from @reduz:
```
3:19 <reduz> Akien: idea is good, but it's not the most efficient way to do it
13:24 <Akien> reduz: any idea on what could be a more efficient implementation?
13:24 <reduz> i don't know, should think about it, the other problem is that doing this also means one need to check for cyclical inclusions so it's not that simple
13:25 <reduz> may be better to ask TMM or someone else for help on this
```

pouleyKetchoupp commented:
I don't mind spending more time to implement this fix in a better way, but if you want to give it to someone who knows the engine better it's fine too :)

Things I would like to clarify:
- Cyclical inclusions: what is this problem exactly? Are we talking about the specific issue of comparing dictionaries, or more generally comparing export variables between instanced scenes and parents?
- More efficient implementation: yes, it's possible to get rid of the allocation/copy of the list of keys/values by having the comparison logic in Dictionary rather than in Variant.
Actually, the `hash` and `duplicate` functions in Dictionary could also get the same optimization.
- Fix == operator: isn't it dangerous in term of functionality or performance to change this operator's behavior to a deep comparison? I would rather make a separate function like `bool compare_values(const Dictionary &p_dictionary)` to be safe.

Xrayez commented:
> This change makes dictionary comparison operators to check for differences in data, following the same logic as for arrays.

If this reflects the same logic as in `Array`, perhaps it could be merged then? This actually solves one of the recent issues described in above PR: #30715.

pouleyKetchoupp committed:
Fixed input events not saved properly when modifying the default actions (regression)

pouleyKetchoupp commented:
I've just pushed a fix for a regression I've spotted.

When modifying inputs for one of the default actions in project settings, it wasn't saved into the project file because no change was detected when comparing dictionaries.

It was due to arrays of events in the input map data being shared between the current settings and the initial settings it was comparing to. So now that deep comparison is done for dictionaries, I had to change project_settings_editor to make new copies of arrays of events when changing some values inside.

QbieShay commented:
Hi, I tested the PR but it doesn't seem to solve the issue for me. Can somebody else test too?

aaronfranke commented:
@pouleyKetchoupp Is this still desired? If so, it needs to be rebased on the latest master branch.

pouleyKetchoupp commented:
Closing in favor of #35816 which addresses the concerns raised on irc about this PR (https://github.com/godotengine/godot/pull/29222#issuecomment-503520898)
